File: nnlist.cpp
#include <iostream>

using namespace std;

int **nnlist(int L, int ndim, int ncoord, int vol){
  
  int i,imod;

  int**  nn = new int*[vol];
  for(i=0 ; i < vol ; i++) nn[i] = new int[ncoord];

  if(ndim == 2 ){
    
    if(ncoord == 4){
      
      for(i=0; i< vol; i++){

	nn[i][0] = (i+1) % L == 0 ? i-L+1 : i+1;
	nn[i][1] = (i < L) ? (vol-L)+i : i-L;
	nn[i][2] = ((i % L) == 0 || i == 0) ? i+L-1 : i-1;
	nn[i][3] = (i >= (vol-L)) ? i-(vol-L) : i+L;
      }

    }else if(ncoord == 6 ){
      
      for(i=0; i<vol; i++){

	nn[i][0]=i+1;
	nn[i][1]=i-L+1;
	nn[i][2]=i-L;
	nn[i][3]=i-1;
	nn[i][4]=i+L-1;
	nn[i][5]=i+L;
	
	if( nn[i][1] <  0  )  nn[i][1]=nn[i][1] + vol;
	if( nn[i][2] <  0  )  nn[i][2]=nn[i][2] + vol;
	if( nn[i][4] >= vol ) nn[i][4]=nn[i][4] - vol;
	if( nn[i][5] >= vol ) nn[i][5]=nn[i][5] - vol; 
	
	imod=(i+1) % L;
	if(imod == 0){
	  nn[i][0]=nn[i][0]-L;
	  nn[i][1]=nn[i][1]-L;
	  if(i == L-1)nn[i][1]=nn[i][1]+vol;
	}
	if(i % L == 0){
	  nn[i][3]=i+L-1;
	  nn[i][4]=nn[i][4]+L;
	  if(i >= vol -L)nn[i][4]=nn[i][4]-vol;
	}
	
      }
      
      
    }
  }
  return nn;
}

File: fortran_version/nnlist.f90
subroutine nnlist(l,ndim,vol,ncoord,nn)

  implicit none

  integer,intent(in)::l,ndim,vol,ncoord
  integer,intent(out),dimension(ncoord,vol)::nn

  integer::i,lm1
  integer::imod

  if(ncoord == 4 .and. ndim == 2)then

     do i=1,vol
        
        nn(1,i)=i+1 !1
        if(mod(i,L) == 0) nn(1,i)=i-L+1
        nn(2,i)=i+L !3
        if(nn(2,i)>vol) nn(2,i)=i-vol+L
        nn(3,i)=i-1 
        if(i == 1 .or. mod(i,L)==1)nn(3,i)=i+(L-1)
        nn(4,i)=i-L !4
        if(i <= L)nn(4,i)=i+vol-L
        
     end do


  else if(ncoord == 6 .and. ndim == 2)then
    
     do i=1,vol
        nn(1,I)=i+1
        nn(2,I)=i+LM1
        nn(3,I)=i+L
        nn(4,I)=i-1
        nn(5,I)=i-LM1
        nn(6,I)=i-L
        
        imod=mod(i,L)   
        if(nn(3,I) > vol)  nn(3,i)=nn(3,i)-vol
        if(nn(2,I) > vol)  nn(2,i)=nn(2,i)-vol
        if(nn(5,I) < 0)   nn(5,i)=nn(5,i)+vol
        if(nn(6,I) < 1)   nn(6,i)=nn(6,i)+vol
        
        if(imod == 1)  then
           nn(4,I)=i+L-1
           nn(5,I)=i-1
        end if


        if(imod == 0) then
           nn(2,I)=I+1
           nn(1,I)=I-L+1
        end if
        
        if(i == 1)  nn(5,i)=vol
        if(i == vol) nn(2,i)=1
        
     end do
     
  end if

end subroutine nnlist
